src/ring_theory/congruence.lean

```diff
@@ -0,0 +1,259 @@
+/-
+Copyright (c) 2022 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+
+import group_theory.congruence
+import algebra.ring.inj_surj
+import algebra.hom.ring
+
+/-!
+# Congruence relations on rings
+
+This file defines congruence relations on rings, which extend `con` and `add_con` on monoids and
+additive monoids.
+
+Most of the time you likely want to use the `ideal.quotient` API that is built on top of this.
+
+## Main Definitions
+
+* `ring_con R`: the type of congruence relations respecting `+` and `*`.
+* `ring_con_gen r`: the inductively defined smallest ring congruence relation containing a given
+  binary relation.
+
+## TODO
+
+* Use this for `ring_quot` too.
+* Copy across more API from `con` and `add_con` in `group_theory/congruence.lean`.
+
+-/
+
+/-- A congruence relation on a type with an addition and multiplication is an equivalence relation
+which preserves both. -/
+structure ring_con (R : Type*) [has_add R] [has_mul R] extends setoid R :=
+(add' : ∀ {w x y z}, r w x → r y z → r (w + y) (x + z))
+(mul' : ∀ {w x y z}, r w x → r y z → r (w * y) (x * z))
```

Comment:
Why don't we extend `add_con` and `con` directly here? It should also allow you to use declarations in the `add_con`/`con` namespace for dot notation.
```suggestion
structure ring_con (R : Type*) [has_add R] [has_mul R] extends
  add_con R, con R
```

Reply:
If I do that then lean forgets that `ring_con` is a structure and `{ ... }` syntax doesn't work
```
set_option old_structure_cmd true

/-- A congruence relation on a type with an addition and multiplication is an equivalence relation
which preserves both. -/
structure ring_con (R : Type*) [has_add R] [has_mul R] extends add_con R, con R.

run_cmd do e ← tactic.get_env, tactic.trace (e.is_structure `ring_con)  -- ff!
```

Reply:
I suppose it's because we're extending a new structure with an `old_structure_cmd`. In any case I can't figure out an easy way to fix this.

Shall we leave a comment for when this hits Lean 4 and we can extend structures without worrying about the newness of our `structure_cmd`s?
```suggestion
-- TODO for mathlib4: define this as `extends add_con R, con R.` instead (this ran into issues
-- that likely derive from `old_structure_cmd`, so should work in Lean 4.)
structure ring_con (R : Type*) [has_add R] [has_mul R] extends setoid R :=
(add' : ∀ {w x y z}, r w x → r y z → r (w + y) (x + z))
(mul' : ∀ {w x y z}, r w x → r y z → r (w * y) (x * z))
```

Reply:
Added something similar.